File: crates/agent-block/tests/fixtures/compile_loop_asr_anthropic_mock.lua
-- Fixture for the compile_loop apply_search_replace (write-channel tool) e2e test.
--
-- Scenario (1 iteration, 2 LLM calls):
--   Call 1: Mock returns two apply_search_replace tool_use blocks (file_a, file_b).
--           compile_loop applies both edits and writes them to disk immediately.
--   Call 2: Mock returns the plain text "DONE" (no SR blocks). Because the edits
--           were applied via the tool channel, compile_loop must proceed to
--           verify instead of treating the missing SR text as a parse failure.
--
-- tool_mode is left unset — the default "auto" must declare apply_search_replace.
--
-- Initial file contents written before the loop:
--   file_a: print("a-old")
--   file_b: print("b-old")
-- After tool-channel apply:
--   file_a: print("a-new")
--   file_b: print("b-new")
-- mock_runner checks that both files contain "new".

local base_url = std.env.get("ANTHROPIC_BASE_URL_TEST")
assert(base_url, "ANTHROPIC_BASE_URL_TEST must be set")

local target_files_env = std.env.get("COMPILE_LOOP_TARGET_FILES")
assert(target_files_env, "COMPILE_LOOP_TARGET_FILES must be set")

local target_files = {}
for p in target_files_env:gmatch("[^:]+") do
    table.insert(target_files, p)
end
assert(#target_files == 2, "expected 2 paths in COMPILE_LOOP_TARGET_FILES, got " .. #target_files)
local file_a_path = target_files[1]
local file_b_path = target_files[2]

do
    local fa = assert(io.open(file_a_path, "w"))
    fa:write('print("a-old")\n')
    fa:close()

    local fb = assert(io.open(file_b_path, "w"))
    fb:write('print("b-old")\n')
    fb:close()
end

local compile_loop = require("compile_loop")

local runner_call_count = 0
local function mock_runner(paths)
    assert(type(paths) == "table", "multi-file mode must pass list to runner, got: " .. type(paths))
    runner_call_count = runner_call_count + 1

    local all_ok = true
    local combined_stdout = ""
    for _, p in ipairs(paths) do
        local f = io.open(p, "r")
        if not f then
            return { ok = false, stderr = "cannot open " .. p, stdout = "", exit_code = 1 }
        end
        local content = f:read("*a") or ""
        f:close()
        combined_stdout = combined_stdout .. content
        if not content:find("new", 1, true) then
            all_ok = false
        end
    end

    return { ok = all_ok, stdout = combined_stdout, stderr = "", exit_code = all_ok and 0 or 1 }
end

local td = compile_loop.make({
    runner = mock_runner,
    edit_mode = "diff",
    llm = {
        provider = "anthropic",
        base_url = base_url,
        api_key = "dummy",
        model = "claude-haiku-mock",
    },
})

local result_json = td.handler({
    spec = "change a-old to a-new and b-old to b-new",
    target_files = { file_a_path, file_b_path },
})

assert(runner_call_count >= 1, "mock_runner must be called at least once, got " .. runner_call_count)

local result = std.json.decode(result_json)
assert(
    result.ok,
    "compile_loop must succeed via the apply_search_replace tool channel, got: " .. (result.summary or "?")
)
assert(result.iters == 1, "loop must converge in 1 iter (tool-channel edits + DONE), got " .. tostring(result.iters))

-- multi-file mode: modified_files must list both tool-written paths.
assert(type(result.modified_files) == "table", "result.modified_files must be a table in multi-file mode")
assert(#result.modified_files == 2, "result.modified_files must contain 2 paths, got " .. #result.modified_files)
assert(result.artifact_path == nil, "result.artifact_path must be nil in multi-file mode")

for _, p in ipairs({ file_a_path, file_b_path }) do
    local f = assert(io.open(p, "r"))
    local content = f:read("*a") or ""
    f:close()
    assert(content:find("new", 1, true), "file " .. p .. " must contain 'new' after tool apply, got: " .. content)
end

print("COMPILE_LOOP_ASR_MOCK_PASS")

File: crates/agent-block/tests/common/compile_loop_asr_anthropic_mock.rs
//! In-process Anthropic Messages API mock server for compile_loop
//! apply_search_replace (write-channel tool) e2e tests.
#![allow(dead_code)]
//!
//! Provides two mock servers:
//!
//! ## Tool round-trip (2-call, 1-iter):
//!   - Call 1: returns two `apply_search_replace` tool_use blocks (file_a and file_b),
//!     stop_reason "tool_use". Also records whether the request declared the
//!     apply_search_replace tool (tool_mode="auto" contract).
//!   - Call 2 (tool_result blocks present): returns the plain text "DONE",
//!     stop_reason "end_turn". compile_loop must proceed to verify with zero
//!     text SR blocks because the edits were applied via the tool channel.
//!
//! ## tool_mode="none" (1-call):
//!   - Records whether the request carried a "tools" key (it must NOT), then
//!     returns path-header SEARCH/REPLACE text for both files in a single turn.
//!
//! Initial file contents written by the Lua fixtures:
//!   file_a: `print("a-old")\n`
//!   file_b: `print("b-old")\n`
//! After successful apply:
//!   file_a: `print("a-new")\n`
//!   file_b: `print("b-new")\n`

use axum::{
    extract::State,
    http::{header, StatusCode},
    response::IntoResponse,
    routing::post,
    Router,
};
use serde_json::json;
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    Arc,
};
use tokio_util::sync::CancellationToken;

/// Shared state for the apply_search_replace mock.
#[derive(Clone)]
pub struct AsrMockState {
    /// Total POST /v1/messages calls.
    pub call_count: Arc<AtomicUsize>,
    /// Calls whose request body declared the apply_search_replace tool.
    pub asr_tool_declared_count: Arc<AtomicUsize>,
    /// Calls whose request body carried any "tools" key at all.
    pub tools_declared_count: Arc<AtomicUsize>,
}

/// Extract absolute file paths from the multi-file lazy-load user message
/// (`Files:\n  <abs_path>\n  <abs_path>`). Returns at most 2 paths.
fn extract_paths_from_request(body: &serde_json::Value) -> Vec<String> {
    let mut paths = Vec::new();
    let messages = match body.get("messages").and_then(|m| m.as_array()) {
        Some(m) => m,
        None => return paths,
    };
    for msg in messages {
        let content = match msg.get("content").and_then(|c| c.as_str()) {
            Some(c) => c,
            None => continue,
        };
        let mut in_files_section = false;
        for line in content.lines() {
            if line.trim() == "Files:" {
                in_files_section = true;
                continue;
            }
            if in_files_section {
                let trimmed = line.trim();
                if trimmed.starts_with('/') {
                    let p = trimmed.to_string();
                    if !paths.contains(&p) {
                        paths.push(p);
                    }
                    if paths.len() >= 2 {
                        return paths;
                    }
                } else if !trimmed.is_empty() {
                    in_files_section = false;
                }
            }
        }
    }
    paths
}

/// True when any user message carries a tool_result content block.
fn has_tool_results(body: &serde_json::Value) -> bool {
    body.get("messages")
        .and_then(|m| m.as_array())
        .map(|msgs| {
            msgs.iter().any(|msg| {
                if msg.get("role").and_then(|r| r.as_str()) != Some("user") {
                    return false;
                }
                msg.get("content")
                    .and_then(|c| c.as_array())
                    .map(|blocks| {
                        blocks
                            .iter()
                            .any(|b| b.get("type").and_then(|t| t.as_str()) == Some("tool_result"))
                    })
                    .unwrap_or(false)
            })
        })
        .unwrap_or(false)
}

/// Record tools-declaration facts about the request into the mock state.
fn record_tools_facts(state: &AsrMockState, body: &serde_json::Value) {
    if let Some(tools) = body.get("tools").and_then(|t| t.as_array()) {
        state.tools_declared_count.fetch_add(1, Ordering::SeqCst);
        let has_asr = tools
            .iter()
            .any(|t| t.get("name").and_then(|n| n.as_str()) == Some("apply_search_replace"));
        if has_asr {
            state.asr_tool_declared_count.fetch_add(1, Ordering::SeqCst);
        }
    }
}

/// POST /v1/messages handler — apply_search_replace tool round-trip.
///
/// Call 1: two apply_search_replace tool_use blocks (one per file).
/// Call 2: plain "DONE" text (edits already applied via the tool channel).
async fn asr_tool_handler(
    State(state): State<AsrMockState>,
    body: axum::body::Bytes,
) -> impl IntoResponse {
    let req_value = match serde_json::from_slice::<serde_json::Value>(&body) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[compile_loop_asr_anthropic_mock] bad request body: {e}");
            let err_body = json!({ "error": format!("bad request: {e}") }).to_string();
            return (
                StatusCode::BAD_REQUEST,
                [(header::CONTENT_TYPE, "application/json")],
                err_body,
            );
        }
    };

    let prev = state.call_count.fetch_add(1, Ordering::SeqCst);
    record_tools_facts(&state, &req_value);

    let response_json = if !has_tool_results(&req_value) {
        // Call 1: emit one apply_search_replace tool_use per file.
        let paths = extract_paths_from_request(&req_value);
        let (path_a, path_b) = if paths.len() >= 2 {
            (paths[0].clone(), paths[1].clone())
        } else {
            ("file_a.lua".to_string(), "file_b.lua".to_string())
        };
        json!({
            "id": format!("msg_asr_turn0_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [
                {
                    "type": "tool_use",
                    "id": "toolu_asr_1",
                    "name": "apply_search_replace",
                    "input": {
                        "path": path_a,
                        "search": "print(\"a-old\")",
                        "replace": "print(\"a-new\")"
                    }
                },
                {
                    "type": "tool_use",
                    "id": "toolu_asr_2",
                    "name": "apply_search_replace",
                    "input": {
                        "path": path_b,
                        "search": "print(\"b-old\")",
                        "replace": "print(\"b-new\")"
                    }
                }
            ],
            "model": "claude-haiku-mock",
            "stop_reason": "tool_use",
            "usage": { "input_tokens": 30, "output_tokens": 25 }
        })
    } else {
        // Call 2: no text SR blocks — the tool channel already applied the edits.
        json!({
            "id": format!("msg_asr_done_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [{ "type": "text", "text": "DONE" }],
            "model": "claude-haiku-mock",
            "stop_reason": "end_turn",
            "usage": { "input_tokens": 40, "output_tokens": 5 }
        })
    };

    (
        StatusCode::OK,
        [(header::CONTENT_TYPE, "application/json")],
        response_json.to_string(),
    )
}

/// POST /v1/messages handler — tool_mode="none" contract.
///
/// Records whether "tools" was declared (must not be), then returns path-header
/// SEARCH/REPLACE text for both files in a single turn.
async fn asr_none_handler(
    State(state): State<AsrMockState>,
    body: axum::body::Bytes,
) -> impl IntoResponse {
    let req_value = match serde_json::from_slice::<serde_json::Value>(&body) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[compile_loop_asr_anthropic_mock/none] bad request body: {e}");
            let err_body = json!({ "error": format!("bad request: {e}") }).to_string();
            return (
                StatusCode::BAD_REQUEST,
                [(header::CONTENT_TYPE, "application/json")],
                err_body,
            );
        }
    };

    let prev = state.call_count.fetch_add(1, Ordering::SeqCst);
    record_tools_facts(&state, &req_value);

    let paths = extract_paths_from_request(&req_value);
    let (path_a, path_b) = if paths.len() >= 2 {
        (paths[0].clone(), paths[1].clone())
    } else {
        ("file_a.lua".to_string(), "file_b.lua".to_string())
    };

    let text = format!(
        "<<< path={path_a} >>>\n<<<<<<< SEARCH\nprint(\"a-old\")\n=======\nprint(\"a-new\")\n>>>>>>> REPLACE\n\n<<< path={path_b} >>>\n<<<<<<< SEARCH\nprint(\"b-old\")\n=======\nprint(\"b-new\")\n>>>>>>> REPLACE"
    );

    let response_json = json!({
        "id": format!("msg_asr_none_{}", prev + 1),
        "type": "message",
        "role": "assistant",
        "content": [{ "type": "text", "text": text }],
        "model": "claude-haiku-mock",
        "stop_reason": "end_turn",
        "usage": { "input_tokens": 20, "output_tokens": 20 }
    });

    (
        StatusCode::OK,
        [(header::CONTENT_TYPE, "application/json")],
        response_json.to_string(),
    )
}

fn new_state() -> AsrMockState {
    AsrMockState {
        call_count: Arc::new(AtomicUsize::new(0)),
        asr_tool_declared_count: Arc::new(AtomicUsize::new(0)),
        tools_declared_count: Arc::new(AtomicUsize::new(0)),
    }
}

async fn spawn(router: Router, bind_msg: &str) -> (String, CancellationToken) {
    let ct = CancellationToken::new();
    let listener = tokio::net::TcpListener::bind("127.0.0.1:0")
        .await
        .unwrap_or_else(|e| panic!("{bind_msg}: {e}"));
    let addr = listener.local_addr().expect("local_addr");
    let ct_shutdown = ct.clone();
    tokio::spawn(async move {
        let _ = axum::serve(listener, router)
            .with_graceful_shutdown(async move { ct_shutdown.cancelled_owned().await })
            .await;
    });
    (format!("http://{addr}"), ct)
}

/// Spawn the apply_search_replace tool round-trip mock.
///
/// # Returns
/// - `base_url`: pass as `ANTHROPIC_BASE_URL_TEST`.
/// - `state`: assert `call_count == 2` and `asr_tool_declared_count >= 1`.
/// - `ct`: call `ct.cancel()` to shut down gracefully.
pub async fn spawn_compile_loop_asr_anthropic_mock_server(
) -> (String, AsrMockState, CancellationToken) {
    let state = new_state();
    let router = Router::new()
        .route("/v1/messages", post(asr_tool_handler))
        .with_state(state.clone());
    let (base_url, ct) = spawn(router, "bind ephemeral port for compile_loop asr mock").await;
    (base_url, state, ct)
}

/// Spawn the tool_mode="none" mock.
///
/// # Returns
/// - `base_url`: pass as `ANTHROPIC_BASE_URL_TEST`.
/// - `state`: assert `tools_declared_count == 0` (no tools were declared).
/// - `ct`: call `ct.cancel()` to shut down gracefully.
pub async fn spawn_compile_loop_tool_mode_none_mock_server(
) -> (String, AsrMockState, CancellationToken) {
    let state = new_state();
    let router = Router::new()
        .route("/v1/messages", post(asr_none_handler))
        .with_state(state.clone());
    let (base_url, ct) = spawn(
        router,
        "bind ephemeral port for compile_loop tool_mode none mock",
    )
    .await;
    (base_url, state, ct)
}
